src/ir/principal-type.cpp

```diff
@@ -0,0 +1,1062 @@
+        return false;
+      }
+    }
+    return true;
+  }
+};
+
+// Assignments for all the kinds of variables that can appear in a principal
+// type.
+struct VarAssignments {
+  AssignmentClasses<VarNullability> nulls;
+  AssignmentClasses<VarExactness> exacts;
+  AssignmentClasses<VarSharedness> shares;
+  AssignmentClasses<VarHeapType> heapTypes;
+  AssignmentClasses<VarType> types;
+
+  bool operator==(const VarAssignments& other) const {
+    return nulls == other.nulls && exacts == other.exacts &&
+           shares == other.shares && heapTypes == other.heapTypes &&
+           types == other.types;
+  }
+
+  bool operator!=(const VarAssignments& other) const {
+    return !(*this == other);
+  }
+};
+
+// apply: replace variables with their assigned values according to
+// `assignments`.
+
+void apply(const VarAssignments& assignments, VarNullability& null) {
+  if (auto* i = std::get_if<Index>(&null)) {
+    null = assignments.nulls.get(*i);
+  }
+}
+
+void apply(const VarAssignments& assignments, VarExactness& exact) {
+  if (auto* i = std::get_if<Index>(&exact)) {
+    exact = assignments.exacts.get(*i);
+  }
+}
+
+void apply(const VarAssignments& assignments, VarSharedness& share) {
+  if (auto* i = std::get_if<Index>(&share)) {
+    share = assignments.shares.get(*i);
+  }
+}
+
+void apply(const VarAssignments& assignments, VarAbsHeapType& type) {
+  apply(assignments, type.share);
+}
+
+void apply(const VarAssignments& assignments, VarDefHeapType& type) {
+  apply(assignments, type.exact);
+}
+
+void apply(const VarAssignments& assignments, VarHeapType& type) {
+  if (auto* i = std::get_if<Index>(&type)) {
+    type = assignments.heapTypes.get(*i);
+  }
+  if (auto* t = std::get_if<VarAbsHeapType>(&type)) {
+    apply(assignments, *t);
+  } else if (auto* t = std::get_if<VarDefHeapType>(&type)) {
+    apply(assignments, *t);
+  }
+}
+
+void apply(const VarAssignments& assignments, VarRef& ref) {
+  apply(assignments, ref.null);
+  apply(assignments, ref.ht);
+}
+
+void apply(const VarAssignments& assignments, VarType& type) {
+  if (auto* i = std::get_if<Index>(&type)) {
+    type = assignments.types.get(*i);
+  }
+  if (auto* t = std::get_if<VarRef>(&type)) {
+    apply(assignments, *t);
+    type = canonicalizeRef(*t);
+  }
+}
+
+void apply(const VarAssignments& assignments, PrincipalType& type) {
+  for (auto& t : type.rparams) {
+    apply(assignments, t);
+  }
+  for (auto& t : type.results) {
+    apply(assignments, t);
+  }
+}
+
+// matchBottom: join the bottom value into the assignemnt in `assignments` for
+// any variables appearing in the second parameter. Returns `true` if this
+// succeeds. TODO: This currently always succeeded, but if we supported repeated
+// variables in outputs properly, we would need to avoid mixing left assignment
+// and right assignemnt, so this would be able to fail.
+
+bool matchBottom(VarAssignments& assignments, const VarNullability& null) {
```

Comment:
I'm confused as to where the word "match" comes from, here and below. Here, can this not be `assignBottom`? Looks like that is what it does, assign the bottom type if the input is an index.